[SDN-1364] Add Network Policy audit logging Enhancement (#617)
openshift/enhancements: enhancements/network/network-policy-audit-logging.md

## Motivation

Many customers require the ability to audit network policy related traffic events for regulatory and security policy compliance. Openshift currently does not have any features that satisfy this requirement. The ACL audit logging will allow customers to monitor NetworkPolicy events and identify patterns involving malicious activity with both allow and deny events. This is necessary in scenarios where
customers require a certain level of compliance, such as monitoring firewall activity, intrusion detection support, or to perform post-mortem analysis.

### Goals

- Activate configurable ACL allow/deny logging in OVN-Kubernetes on a per namespace basis.
- Allow the cluster administrator configure the auditing via the Cluster Network Operator's configuration.
- Allow the cluster administrator set the logging level via a namespace's `.metadata.annotations` field.
- Collect the relevant data from the ovn-controller logs and present it to the cluster administrator.  
- Alow the cluster administrator to extract the logs with an `oc adm node-logs --path=ovn/` command

### Non-Goals

> 

Author: Does not apply to SDN

---
title: audit-logging-of-network-policy-events
authors:
  - "@astoycos"
reviewers:
  - "@abhat" 
  - "@vpickard"
  - "@trozet" 
  - "@Billy99"
approvers:
  - "@knobunc"
creation-date: 2020-12-11
last-updated: 2020-12-11
status: implementable

---

# Audit Loggging of Network Policy Events 

## Release Signoff Checklist

> - [ ] Enhancement is `implementable`

Reviewer: You have it marked implementable above.  In general, update this checklist as appropriate

- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> The OVN-Kubernetes CNI uses [OVN](https://www.ovn.org) to implement node overlay networks for Kubenetes. When Ovn-Kubernetes is used as the CNI for an openshift cluster OVN ACLs are the used to implement Kubernetes NetworkPolicies.  ACL's can either allow or deny traffic by matching on packets with specific rules.  Built into the OVN ACL feature is the ability to specify logging for each "allow" or "deny" rule.  The enhancement will activate this logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used, therefore showing valiuable realtime information involving NetworkPolicies 

Reviewer: ```suggestion
The OVN-Kubernetes network type uses [OVN](https://www.ovn.org) to implement node overlay networks for Kubernetes. When OVN-Kubernetes is used as the network type for an OpenShift cluster, OVN ACLs are used to implement Kubernetes network policies (`NetworkPolicy` resources).  ACLs can either allow or deny traffic by matching on packets with specific rules.  Built into the OVN ACL feature is the ability to specify logging for each "allow" or "deny" rule.  This enhancement will activate this logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used. These logs will show valuable real-time information involving network policies. 
```

Reviewer: I'd run a spell check on the doc

> The OVN-Kubernetes CNI uses [OVN](https://www.ovn.org) to implement node overlay networks for Kubenetes. When Ovn-Kubernetes is used as the CNI for an openshift cluster OVN ACLs are the used to implement Kubernetes NetworkPolicies.  ACL's can either allow or deny traffic by matching on packets with specific rules.  Built into the OVN ACL feature is the ability to specify logging for each "allow" or "deny" rule.  The enhancement will activate this logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used, therefore showing valiuable realtime information involving NetworkPolicies 

Author: Yeah based on the other enhancements I think Openshift should be capitalized

> The OVN-Kubernetes CNI uses [OVN](https://www.ovn.org) to implement node overlay networks for Kubenetes. When Ovn-Kubernetes is used as the CNI for an openshift cluster OVN ACLs are the used to implement Kubernetes NetworkPolicies.  ACL's can either allow or deny traffic by matching on packets with specific rules.  Built into the OVN ACL feature is the ability to specify logging for each "allow" or "deny" rule.  The enhancement will activate this logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used, therefore showing valiuable realtime information involving NetworkPolicies 

Author: Thanks for all the Grammar tips... never was my favorite subject

## Motivation

> Currently openshift has no ability to audit security events reguarding NetworkPolicies, and there are many customers which require this feature for security policy and regulation. The logging will allow customers to monitor NetworkPolicy events and identify pattterns involving malicious activity with both allow and deny events. This is necessary in scenarios where customers require a certain level of compliance, such as monitoring firewall activity, intrusion detection support, or fo post-mortem analysis 

Reviewer: ```suggestion
Currently OpenShift administrators have no ability to audit security events regarding network policies. There are many customers which require this feature for security policy and regulation. The logging will allow customers to monitor `NetworkPolicy` events and identify patterns involving malicious activity with both allow and deny events. This is necessary in scenarios where customers require a certain level of compliance, such as monitoring firewall activity, intrusion detection support, or for post-mortem analysis. 
```

### Goals

- Activate configurable ACL allow/deny logging in Ovn-Kubernetes on a per namespace basis 

> - Allow the Customer to control the logging rate for this feature via the Cluster Network Operator's config 

Author: Agreed

> - Allow the Customer to configure this fetaure's use and level via a Namespace's `.metadata.annotations` field 

Reviewer: On these last two, I don't think you need to include mechanism in the goal.  The goals could be:

- Allow cluster administrators to set a global logging level and rate limit

Presumably this *could* be on a per-namespace basis too, right?  What's your thinking on which requirements are on a per namespace basis vs global configuration?

Author: Based on the current implementation Logging level can be set on a per namespace basis while the rate limit variable will need to be set globally

> - Collect the relevant data from the ovn-controller logs and preset it to the Customer in an seamless manner 

Reviewer: ```suggestion
- Collect the relevant data from the `ovn-controller` logs and present it to the cluster administrator 
```

I like using "cluster administrator" here as it more clearly identifies the target persona we're looking to support with this feature.

### Non-Goals

- Network Policy Object Logging (`oc describe <entitiy>` takes care of that)

> ## Proposal

Reviewer: The proposal doesn't include anything about setting the log level, or how this is enabled on a per-namespace basis.  (though I know it's still marked WIP)

Author: Good point I missed that :)

> The Enhancment's implementation starts with somce changes to the [network.operator.openshift.io](https://github.com/openshift/api/blob/master/operator/v1/types_network.go) API.  These changes involve adding one optional flag to the `OVNKubernetesConfig` struct as follows. 

Reviewer: Can you call out specifically which field is new in the text here?

Author: Yep!

```go
// ovnKubernetesConfig contains the configuration parameters for networks
// using the ovn-kubernetes network project
type OVNKubernetesConfig struct {
	// mtu is the MTU to use for the tunnel interface. This must be 100
	// bytes smaller than the uplink mtu.
	// Default is 1400
	// +kubebuilder:validation:Minimum=0
	// +optional
	MTU *uint32 `json:"mtu,omitempty"`
	// geneve port is the UDP port to be used by geneve encapulation.
	// Default is 6081
	// +kubebuilder:validation:Minimum=1
	// +optional
	GenevePort *uint32 `json:"genevePort,omitempty"`
	// HybridOverlayConfig configures an additional overlay network for peers that are
	// not using OVN.
	// +optional
	HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
	// ipsecConfig enables and configures IPsec for pods on the pod network within the
	// cluster.
	// +optional
	IPsecConfig *IPsecConfig `json:"ipsecConfig,omitempty"`

>   // aclLoggingRateLimit allows for the configuring of max acl logging rate 

Reviewer: What's the unit for the rate?

Author: messages per second

  // +optional

>   aclLoggingRateLimit  *uint32 `json:"aclLoggingRateLimit,omitempty"`

Reviewer: There's an indentation issue here (tabs vs spaces)

```

> The flag `aclLoggingRateLimit` defaults to `20`. 

Reviewer 2: What does`20` imply? 20 logs per second? First 20 logs are written and the rest dropped? Does it tell you when logs are dropped or how many are dropped?

Author: 20 messages per second, and yeah once that meter argument is surpassed any additional logs are dropped. 

```
 The  action  argument specifies what should happen
 when this meter is exceeded. The only supported action is drop.
```

> The flag `aclLoggingRateLimit` defaults to `20`. 

Reviewer: 20 what?  (units)

Author: messages per second

This will result in an updated OVNKubernetes Operator configuration object: (.spec.defaultNetwork.ovnKubernetesConfig) which will enable cluster-wide configuration of acl logging at cluster installation time: 

```go
spec:
  defaultNetwork:
    type: OVNKubernetes
    ovnKubernetesConfig:
      mtu: 1400
      genevePort: 6081
      ipsecConfig: {}
      aclLoggingRateLimit: 20
```

### User Stories

> Detail the things that people will be able to do if this is implemented.

Reviewer 2: Remove the document template directions.

Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.

Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.

#### Story 1

> As an Openshift user I want to see what Network Policies and accompanied ACLs are dropping the most traffic in a cluster. For example, when a packet runs into a `drop` ACL it will log a message similar to the following 

Reviewer 2: - Should be consistent with Openshift vs OpenShift and Network Policies vs NetworkPolicies (global comment).
- End the last sentence with a colon.

```
2021-01-05T17:34:02.675Z|00004|acl_log(ovn_pinctrl0)|INFO|name="<Network Policy Name>", verdict=drop, severity=info: icmp,vlan_tci=0x0000,dl_src=50:54:00:00:00:02,dl_dst=50:54:00:00:00:01,nw_src=192.168.0.3,nw_dst=192.168.0.2,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0
```

#### Story 2

> As an Openshift user I want to control the rate of ACL logging to ensure I can extract the useful information invloving accept/reject actions in high hit rate scenarios. 

Reviewer 2: invloving -> involving

#### Story 3 

As an Openshift user I want to monitor in realtime traffic that is hitting any network policies I have implemented. 

> --- TBD in implementation (Realtime Monitering would be nice here)

Reviewer 2: Monitering -> Monitoring

### Implementation Details/Notes/Constraints 

> What are the caveats to the implementation? What are some important details that

Reviewer 2: Remove the document template directions.

didn't come across above. Go in to as much detail as necessary here. This might
be a good place to talk about core concepts and how they relate.

> The following codebase changes will need to done before this feature becomes usable  

Reviewer 2: End the sentence with a colon.

- OVN Kubernetes 

>   - Move from a cluster wide namespace default deny port group to namespace specific one 

Reviewer 2: Period

>   - Add logging flag and custom OVN meter to all created ACLs, see below 

Reviewer 2: -"... to all created ACLs, see below" ->  "... to all created ACLs. See below see below for an example of additional logging needed:"
- How many spots in the code will need this? Ballpark: 2, 50, 100, more?

Author: The Upstream work has merged -> https://github.com/ovn-org/ovn-kubernetes/pull/1654 The change was only needed in around 10 places in the code base

  ```
  fmt.Sprintf("log=%v", aclLogging != ""),
  fmt.Sprintf("severity=%s", getACLLoggingSeverity(aclLogging)), "meter=acl-logging",
  ```
- Cluster Network Operator 
  
  - Digest the `aclLoggingRateLimit` Flag and start OVN-K with the configured value 

- Logging Stack 
  
  - We need a way to collect and present ACL logging from the OVN controller logs 

### Risks and Mitigations

> What are the risks of this proposal and how do we mitigate. Think broadly. For

Reviewer 2: Remove the document template directions.

example, consider both security and how this will impact the larger OKD
ecosystem.

How will security be reviewed and by whom? How will UX be reviewed and by whom?

Consider including folks that also work outside your immediate sub-project.

> - How do we control who is able to see the audit logs. 

Reviewer 2: audit logs. -> audit logs?

   - Only Admins should be able to see the audit logs since it could help expose holes in a cluster's network policy structure 

- Scale questions involving the addition of OVN meters to all created ACL's 

>   - Adding meters to each ACL does add a slightly larger overhead in OVN

Reviewer 2: These two bullets could probably be one:
"Adding meters to each ACL does add a slightly larger overhead in OVN, but based on early upstream testing with logging both enabled and disabled, there were negligible performance impact detected."

  - Based on Early upstream testing with logging both enabled and disabled there were negligable performance changes  

## Design Details

### Open Questions 

1. How should the ACL logs be digested from the ovn-controller logs and presented to the user? 

>   - Currenlty in OCP API audit logging is dumped directly onto the Nodes, should ACL logging do the same? 

Reviewer 2: - Currenlty -> Currently
- Break into two sentences: Currently in OCP, API audit logging is dumped directly onto the Nodes. Should ACL logging do the same?

>   - We could leverage OCP's logging stack to pull the logs and present to the user with a Kiabana Dashboard, However the logging 

Reviewer 2: - Remove the "We could ..." and change to: OCP's logging stack could be used to pull the logs and present them to the user with a Kiabana Dashboard. However the logging stack is an optional feature in an OCP cluster so this feature would require the logging stack to work properly.

>           stack is an optional feature in an OCP cluster 

Reviewer: Agree with the open question. We should be clear about two cases: with cluster logging enabled, and what experience is possible with what's enabled by default.

We may want to go ahead and get a change into OVN to get these logs in their own log file. Can you add that to this enhancement? Leaving it in the ovn-controller log file seems pretty messy.

Author: ACK Adrian on the OVN team is aware of the need to push the audit logs to their own file, I am staying in touch with him to track the work.

Author: Yes, I have already spoken with Adrian from the OVN team regarding a separate log file for this feature

> 2. Based on the solution to question 1., Should a must-gather collect these logs, or should they be accessible only at runtime? 

Reviewer 2: Change to:
2. Based on the solution to question 1:
   - Should a must-gather collect these logs, or should they be accessible only at runtime?

---
title: audit-logging-of-network-policy-events
authors:
  - "@astoycos"
reviewers:
  - "@abhat" 
  - "@vpickard"
  - "@trozet" 
  - "@Billy99"
approvers:
  - "@knobunc"
creation-date: 2020-12-11
last-updated: 2020-12-11
status: implementable

---

# Audit Loggging of Network Policy Events

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> The OVN-Kubernetes network type uses [OVN](https://www.ovn.org) to implement node overlay networks for Kubernetes. When OVN-Kubernetes is used as the network type for an Openshift cluster, OVN ACLs are the used to implement Kubernetes  network policies (`NetworkPolicy` resources).  ACL's can either allow or deny traffic by matching on packets with specific rules. Built into the OVN ACL feature is

Reviewer 3: `OVN ACLs are the used` I'd rather say  `are used` or `are the ones used`

Reviewer 3: `Kubernetes'  network policies`

the ability to specify logging for each "allow" or "deny" rule.  This enhancement will activate the OVN ACL feature logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used, thereby showing valuable realtime information involving network policies.

## Motivation

Many customers require the ability to audit network policy related traffic events for regulatory and security policy compliance. Openshift currently does not have any features that satisfy this requirement. The logging will allow customers to monitor NetworkPolicy events and identify patterns involving malicious activity with both allow and deny events. This is necessary in scenarios where
customers require a certain level of compliance, such as monitoring firewall activity, intrusion detection support, or to perform post-mortem analysis.

> ### Goals

Reviewer: I think we should also allow opting in to allow and/or deny events.

Reviewer: i see this got added

```go
// ovnKubernetesConfig contains the configuration parameters for networks
// using the ovn-kubernetes network project
type OVNKubernetesConfig struct {
 // mtu is the MTU to use for the tunnel interface. This must be 100
 // bytes smaller than the uplink mtu.
 // Default is 1400
 // +kubebuilder:validation:Minimum=0
 // +optional
 MTU *uint32 `json:"mtu,omitempty"`
 // geneve port is the UDP port to be used by geneve encapulation.
 // Default is 6081
 // +kubebuilder:validation:Minimum=1
 // +optional
 GenevePort *uint32 `json:"genevePort,omitempty"`
 // HybridOverlayConfig configures an additional overlay network for peers that are
 // not using OVN.
 // +optional
 HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
 // ipsecConfig enables and configures IPsec for pods on the pod network within the
 // cluster.
 // +optional
 IPsecConfig *IPsecConfig `json:"ipsecConfig,omitempty"`
 
 <--- BEGIN NEW CODE --->

>  // aclLoggingRateLimit allows for the configuring of max acl logging rate

Reviewer 4: I strongly suspect we're going to need to add more auditing features. I'd suggest moving this "down a level" by adding an intermediate struct.

Something like
```go
type OVNKubernetesConfig struct {

    // PolicyAuditing is the configuration for network 
    PolicyAuditing *PolicyAuditConfig
}

type PolicyAuditConfig {

    // RateLimit is the approximate maximum number of messages to send per-second per-node. If
    // unset, no limit is applied.
    RateLimit *uint32

    // Destination is the destination for policy log messages. Messages are output in syslog format
    // Valid values are:
    // - "stdout" or "stderr: the audit logging container outputs to its stdout or stderr
    // - "udp:host:port" for sending syslog over UDP
    // - "tcp:host:port" for sending syslog over TCP
    Destination string

    // SyslogFacility the RFC5424 facility for generated messages, e.g. "kern". Default is "local0"
    SyslogFacility *string

    // DefaultLogLevelAllow is the syslog severity used for logging allow events.
    // Allow events are generated whenever a packet is accepted by a network policy, subject
    // to rate limiting. If unset, "allow" messages are not sent. This can also be configured on a
    // per-namespace basis.
    // acceptable values are emer, err, warn, info, or dbg
    DefaultLogLevelAllow *string

    // DefaultLogLevelDeny is the syslog severity used for logging deny events.
    // Allow events are generated whenever a packet is accepted by a network policy, subject
    // to rate limiting. If unset, "deny" messages are not sent. This can also be configured on a
    // per-namespace basis.
    // acceptable values are emer, err, warn, info, or dbg
    DefaultLogLevelDeny *LogLevel
}
```

**Note:** There is a mismatch between ovn severity and syslog severity. Maybe we can clean that up?

Reviewer 4: We may also wish to enable ACL logging directly to the host's journal if there is rate limiting applied. @smarterclayton tends to have concerns about journald throughput, but in this case, we can cap the load.

Author: Agreed I really like this idea, would we allow host journal logging in parallel with logging to another destination? Or can we just have one of the destinations specified as `host`? (And if the `host` destination is chosen some default rate limit is applied)

Author: The only other note here is that I think for now the `DefaultLogLevelAllow` and `DefaultLogLevel` should not be included here since currently the CNO has no control over OVN-K's internal ACL logging levels... However this could be something we decide to implement in the future.

}
```

The flag `aclLoggingRateLimit` defaults to `20` messages per second.

This will result in an updated OVNKubernetes Operator configuration object: `(.spec.defaultNetwork.ovnKubernetesConfig)` which will enable cluster-wide configuration of ACL logging at cluster installation time:

```go
spec:
  defaultNetwork:
    type: OVNKubernetes
    ovnKubernetesConfig:
      mtu: 1400
      genevePort: 6081
      ipsecConfig: {}
      aclLoggingRateLimit: 20
```

To enable the ACL logging on a per namespace basis using metadata annotations in the namespace's definition like the following

```go
kind: Namespace
apiVersion: v1
metadata:
  name: tenantA
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "notice" }'
```

or

```go
kind: Namespace
apiVersion: v1
metadata:
  name: tenantB
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "notice" }'
```

The logging can be activated for either allow, drop, or allow and drop actions. The severity must be one of `alert`, `warning`,
`notice`, `info`, or `debug` as described in [OVN documentation](http://www.openvswitch.org/support/dist-docs/ovn-nbctl.8.html).

each level is listed in order or descending severity

> ```go

Reviewer 4: These values aren't super interesting, since that's just an ovn convention. The severity is just plumbed through from the vswitch to the output log. I would remove this section.

Author: Sounds good to me, I will just say "Refer to ovs documentation" or something along those lines

Reviewer 4: No, that doesn't matter; we generate the severity. It would be something consumed by the log aggregator.

alert    A major failure forced a process to abort.

warning  A high-level operation or a subsystem failed.  Attention
         is warranted.

notice   A low-level operation failed, but higher-level subsystems
         may be able to recover.

info     Information that may be useful in retrospect when
         investigating a problem.

debug    Information useful only to someone with intricate
         knowledge of the system, or that would commonly cause too-
         voluminous log output.  Log messages at this level are not
         logged by default.
```

> ### User Stories

Reviewer: Let's add something about how these logs are accessed with what's included in OpenShift by default - presumably following the same pattern as the API audit log

Author: Added a user story involving mimicking the behavior of the `oc adm` command used to see API audit logs located on nodes.

#### Story 1

As an Openshift user I want to see what Network Policies and accompanied ACLs are dropping the most traffic in a cluster. For example, when a packet runs into a `drop` ACL it will log a message similar to the following:

>  ```go

Reviewer: nit: not actually golang ... maybe sh?  I'm not sure what is best for quoting a log message ...

@dhellmann any ideas for what's appropriate here since you did the cleanup for turning on linting?

>   2021-01-05T17:34:02.675Z|00004|acl_log(ovn_pinctrl0)|INFO|name="<Network Policy Name>", verdict=drop, severity=info: icmp,vlan_tci=0x0000,dl_src=50:54:00:00:00:02,dl_dst=50:54:00:00:00:01,nw_src=192.168.0.3,nw_dst=192.168.0.2,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0

Reviewer 5: For user to be able to make sense of this, the source/destination podname and namespace would need to also be included, otherwise they'll need to manually resolve the IP which is potentially valid for a short period of time, which IMO would limit the usefulness of this feature.

As a user I'd like to see something like: "NetworkPolicy namespace1/X is dropping icmp/8 traffic from namespace2/mypod-abcdef"

Why? If NetworkPolicies are dropping traffic it probably means my app is probably broken in some way, but that might be non-obvious. May help with #613 

I think it should work in a similar way to pod events, eg. when waiting for PVC's to become available - it'll only log a message every ~minute and the Kubernetes will fold these into a single message and say it's been repeated n times.

Author: Hey @jackevans43 Thanks for the review! I completely agree with you on this, however I think it may be out of the scope of the enhancement since it would require quite a bit more infrastructure to resolve the IPs to pod names and ACLs directly to Network Policies, Etc. However, I fully expect there to be follow up epics regarding network policy observability based on the overall observability epic spearheaded by @russellb

Reviewer: I totally agree with the improvement, and that it'd be a future step.  I think it would be good to document this somewhere as one of the next steps with this feature.  Maybe we could have a "future work" section if there's not already something like that in the template.  I would also add integration with the OpenShift console in that "future work" section.

Reviewer 4: We should try and generate an IP audit log in ovn-kube master, since we control IPAM there.

Reviewer: That's a good idea.  It's probably already in the northd log, but of course that's not persisted (unless you use cluster logging)

 ```

#### Story 2

As an Openshift user I want to control the rate of ACL logging to ensure I can extract the useful information involving accept/reject actions in high hit rate scenarios.

#### Story 3

As an Openshift user I want to monitor in realtime traffic that is hitting any network policies I have implemented.

> --- TBD in implementation (Realtime Monitoring would be nice here)

Reviewer 3: Does it make sense to include this TBD around here? I mean, this is just an EP, so implementation will obviously follow-up on what you propose here ;)

> ### Implementation Details/Notes/Constraints

Reviewer: might need work in OVN to get the log solmewhere else?  or work in ovn-kube to get the data somewhere else ...

Author: I am working to get the ACL logging into it's own file

Reviewer: any tracker for this you could link to?  then we could resolve it as an open item

Author: So after a week messing around with trying to use OVS's `syslog` mechanism + a sidecar container I think we will just be able to use native OVN to export only the ACL logs to a file using the `ovn-controller's `--logfile flag` (Since currently OCP does not even use that feature) I.e we shouldn't need any OVN changes

The following codebase changes will need to done before this feature becomes usable:  

- OVN Kubernetes

  - Move from a cluster wide namespace default deny port group to namespace specific ones.
  - Add logging flag and custom OVN meter to all created ACLs:

   ```go
    fmt.Sprintf("log=%v", aclLogging != ""),
    fmt.Sprintf("severity=%s", getACLLoggingSeverity(aclLogging)), "meter=acl-logging",
   ```

  - Upstream work

- Cluster Network Operator
  
  - Digest the `aclLoggingRateLimit` Flag and start OVN-K with the configured value.

- Logging Stack
  

>   - We need a way to collect and present ACL logging from the OVN controller logs.

Reviewer: I think we can clarity this as a demo and documentation task.  I don't think any changes are required.  We just want to demonstrate the use case.

Author: ACK agreed

  versions - i.e. `x.y.N->x.y.N+2` should work without requiring `x.y.N->x.y.N+1`
  as an intermediate step.
- Minor version upgrades - you only need to support `x.N->x.N+1` upgrade
  steps. So, for example, it is acceptable to require a user running 4.3 to
  upgrade to 4.5 with a `4.3->4.4` step followed by a `4.4->4.5` step.
- While an upgrade is in progress, new component versions should
  continue to operate correctly in concert with older component
  versions (aka "version skew"). For example, if a node is down, and
  an operator is rolling out a daemonset, the old and new daemonset
  pods must continue to work correctly even while the cluster remains
  in this partially upgraded state for some time.

Downgrade expectations:

- If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
  misbehaving, it should be possible for the user to rollback to `N`. It is
  acceptable to require some documented manual steps in order to fully restore
  the downgraded cluster to its previous state. Examples of acceptable steps
  include:
  - Deleting any CVO-managed resources added by the new version. The
    CVO does not currently delete resources that no longer exist in

>     the target version.

Reviewer: you've got a bunch of boilerplate text near the bottom of the doc left over from the template that you can remove

Author: - I will change it to provisional 

- The Jira issues are generally tracking all the implementable work so I can just link to those 

- Yep I wasn't sure If I was going to need to fill all that in at some point, I will clean it up
